library/nsxt_upgrade_run.py

```diff
@@ -0,0 +1,270 @@
+#!/usr/bin/env python
+#
+# Copyright 2018 VMware, Inc.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,
+# BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: nsxt_upgrade_run
+short_description: 'Start the upgrade'
+description: 'Upgrade will start as per the upgrade plan.'
+version_added: '2.7'
+author: 'Rahul Raghuvanshi'
+options:
+    hostname:
+        description: 'Deployed NSX manager hostname.'
+        required: true
+        type: str
+    username:
+        description: 'The username to authenticate with the NSX manager.'
+        required: true
+        type: str
+    password:
+        description: 'The password to authenticate with the NSX manager.'
+        required: true
+        type: str
+    paused_upgrade:
+        description: 'Mode of upgrade'
+        required: true
+        type: bool
+'''
+
+EXAMPLES = '''
+- name: Runs the upgrade
+  nsxt_upgrade_run:
+      hostname: "10.192.167.137"
+      username: "admin"
+      password: "Admin!23Admin"
+      validate_certs: False
+      paused_upgrade: True
+'''
+
+RETURN = '''# '''
+
+import json, time
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.vmware_nsxt import vmware_argument_spec, request
+from ansible.module_utils.common_utils import get_attribute_from_endpoint, clean_and_get_params, get_upgrade_orchestrator_node
+from ansible.module_utils._text import to_native
+
+def get_upgrade_status(module, manager_url, mgr_username, mgr_password, validate_certs):
+  '''
+  Get the current status of upgrade at the start.
+  Doesn't upgrade if any component is in progress 
+  or system is already upgraded.
+  '''
+  upgrade_status = get_attribute_from_endpoint(module, manager_url, '/upgrade/summary',
+                        mgr_username, mgr_password, validate_certs, 'upgrade_status', 
+                        False)
+
+  if upgrade_status == 'IN_PROGRESS' or upgrade_status == 'PAUSING':
```

Comment:
what about failure

Reply:
If it in failed state, still we need to run.